test/common/http/filter/ext_authz_test.cc

```diff
@@ -0,0 +1,312 @@
+namespace ExtAuthz {
+
+class HttpExtAuthzFilterTestBase {
+public:
+  HttpExtAuthzFilterTestBase() {}
+
+  FilterConfigSharedPtr config_;
+  Envoy::ExtAuthz::MockClient* client_;
+  std::unique_ptr<Filter> filter_;
+  NiceMock<MockStreamDecoderFilterCallbacks> filter_callbacks_;
+  Envoy::ExtAuthz::RequestCallbacks* request_callbacks_{};
+  TestHeaderMapImpl request_headers_;
+  Buffer::OwnedImpl data_;
+  Stats::IsolatedStoreImpl stats_store_;
+  NiceMock<Runtime::MockLoader> runtime_;
+  NiceMock<Upstream::MockClusterManager> cm_;
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  Network::Address::InstanceConstSharedPtr addr_;
+  NiceMock<Envoy::Network::MockConnection> connection_;
+};
+
+class HttpExtAuthzFilterTest : public testing::Test, public HttpExtAuthzFilterTestBase {
+public:
+  HttpExtAuthzFilterTest() {}
+
+  void SetUpTest(const std::string json) {
+    envoy::config::filter::http::ext_authz::v2::ExtAuthz proto_config{};
+    MessageUtil::loadFromJson(json, proto_config);
+    config_.reset(new FilterConfig(proto_config, local_info_, stats_store_, runtime_, cm_));
+
+    client_ = new Envoy::ExtAuthz::MockClient();
+    filter_.reset(new Filter(config_, Envoy::ExtAuthz::ClientPtr{client_}));
+    filter_->setDecoderFilterCallbacks(filter_callbacks_);
+    addr_ = std::make_shared<Network::Address::Ipv4Instance>("1.2.3.4", 1111);
+  }
+
+  const std::string filter_config_ = R"EOF(
+    {
+      "grpc_service": {
+          "envoy_grpc": { "cluster_name": "ext_authz_server" }
+      },
+      "failure_mode_allow": true
+    }
+  )EOF";
+};
+
+typedef envoy::config::filter::http::ext_authz::v2::ExtAuthz CreateFilterConfigFunc();
+
+class HttpExtAuthzFilterParamTest : public TestWithParam<CreateFilterConfigFunc*>,
+                                    public HttpExtAuthzFilterTestBase {
+public:
+  virtual void SetUp() {
+    envoy::config::filter::http::ext_authz::v2::ExtAuthz proto_config = (*GetParam())();
+    config_.reset(new FilterConfig(proto_config, local_info_, stats_store_, runtime_, cm_));
+
+    client_ = new Envoy::ExtAuthz::MockClient();
+    filter_.reset(new Filter(config_, Envoy::ExtAuthz::ClientPtr{client_}));
+    filter_->setDecoderFilterCallbacks(filter_callbacks_);
+    addr_ = std::make_shared<Network::Address::Ipv4Instance>("1.2.3.4", 1111);
+  }
+};
+
+template <bool failure_mode_allow_value>
+envoy::config::filter::http::ext_authz::v2::ExtAuthz GetFilterConfig() {
+  const std::string json = R"EOF(
+    {
+      "grpc_service": {
+          "envoy_grpc": { "cluster_name": "ext_authz_server" }
+      },
+    }
+  )EOF";
+  envoy::config::filter::http::ext_authz::v2::ExtAuthz proto_config{};
+  MessageUtil::loadFromJson(json, proto_config);
+  proto_config.set_failure_mode_allow(failure_mode_allow_value);
+  return proto_config;
+}
+
+INSTANTIATE_TEST_CASE_P(ParameterizedFilterConfig, HttpExtAuthzFilterParamTest,
+                        Values(&GetFilterConfig<true>, &GetFilterConfig<false>));
```

Comment:
It's probably OK, but I'm not a huge fan of this kind of cross-product testing in unit tests. Ideally each unit test stands alone and tests something discrete, validating the entire config cross-product via the same set of tests will result in combinatorial explosion.

Reply:
The first cut of the pr had single tests which covered `fail_mode_allow=true` and specific tests for `fail_mode_allow=false` case. One of @ggreenway  review was to create broader coverage with parameter based tests.
I don't have a preference and would be ok with changing it to single tests. As long as there is consensus as to what the vote is.

Reply:
It's fine what you have here, just wanted to understand why this was chosen.